API/ev3_timer.c

```diff
@@ -12,268 +16,552 @@
+            break;
+
+        // and handle it
+        if (!tmThreadProcessEvent(&event, &localCopy))
+            break;
+    }
+
+    // clean up the shadow array
+    tmArrayFree(&localCopy);
+    return NULL;
+}
+
+void tmThreadBlockSignals()
+{
+    // block all signals to prevent interruptions, delays etc.
+    sigset_t signalMask;
+    sigfillset(&signalMask);
+    pthread_sigmask(SIG_BLOCK, &signalMask, NULL);
 }
 
-unsigned long TimerMS(uint8_t Timer)
+bool tmThreadWaitForEvent(struct epoll_event *output)
+{
+    while (true)
+    {
+        // let's try to wait
+        int result = epoll_wait(tcb.waiterEpollFd, output, 1, -1);
+        if (result > 0) // event read successfully
+            return true;
+        else if (result == 0) // no event read (shouldn't happen, but let's handle it)
+            continue;
+        else // an error happened
+        {
+            if (errno == EINTR) // signal arrived (shouldn't happen, but let's handle it)
+                continue;
+
+            // we don't know how to handle other errors, let's fail
+            REPORT_AND_RETURN("epoll_wait() failed", false);
+            return false;
+        }
+    }
+}
+
+bool tmThreadProcessEvent(struct epoll_event *pEvent, CallbackBlock *pLocalCopy)
+{
+    // the termination eventfd is tagged with TERMINATOR
+    // when we receive an event from it, quit the loop
+    if (pEvent->data.u32 == TERMINATOR)
+        return false;
+    // otherwise the tag is the interval enum value
+    TimerInterval interval = pEvent->data.u32;
+
+    // critical section
+    LOCK_OR_RETURN(waiter, false);
+    // mirror the corresponding callback block
+    // (this avoids the need for recursive mutexes)
+    bool allocOk = tmArrayMirror(pLocalCopy, &tcb.callbacksFor[interval]);
+    // leave critical section
+    UNLOCK(waiter);
+
+    if (!allocOk)
+        return false;
+
+    // un-flag the FD as readable
+    // also get the elapsed expiration count to loop multiple times, if needed
+    uint64_t elapsedExpirations = 0;
+    int result = read(pLocalCopy->timer_fd,
+                      &elapsedExpirations,
+                      sizeof(elapsedExpirations));
+    RETURN_ON_FAIL(result, "read(timerfd) failed", false);
+
+    // and finally, call the function the needed number of times
+    for (uint64_t n = 0; n < elapsedExpirations; n++)
+        tmArrayCallAll(pLocalCopy, interval);
+
+    return true;
+}
+
+
+void tmArrayCallAll(CallbackBlock *pBlock, TimerInterval interval)
+{
+    // prevent SIGSEGV
+    if (pBlock->elements == NULL)
+        return;
+
+    // call all the functions
+    for (size_t i = 0; i < pBlock->count; i++)
+        pBlock->elements[i](interval);
+}
+
+// DESTRUCTION //
+
+bool TimerExit()
 {
-  if (Timer < NUM_MS_TIMERS)
-	return TimerGetMS() - msTimers[Timer];
-  return 0;
+    bool success = true;
+
+    if (!TimerInitialized())
+        return true;
+
+    // stop the waiter thread
+    success &= tmExitThread();

```

Comment:
`&&=`?

Reply:
If I understand it correctly, that would work similar to returning false directly. I wanted to always deinitialize everything, even if a preceeding deinitialization section failed. Using an if and setting a variable to false will be more readable though.

Reply:
Ah, right. Shortcircuiting! I think too that `if` clauses are probably better ^^'.

Reply:
Resolved in c83c98612da64672bfa2669ca0e6c05f9aab516d :+1: